### ISSUE TITLE
RUFF012: False positive when using inheritance 

### ISSUE THREAD

**Issue by frenck** [astral-sh/ruff#5429]
ruff version: 0.0.275

I'm running into a false positive, as it ignores inheritance.

Example:

```py
class MyBaseClass:

    example: ClassVar[set[str]]

class NotMyBaseClass(MyBaseClass):

    example = {"O", "M", "G", "Puppies!"}
```

RUF012 will now trigger on the `NotMyBaseClass.example`: ```RUF012 Mutable class attributes should be annotated with `typing.ClassVar` ```

However, this is already the case...

../Frenck

**Comment by charliermarsh** [comment]
Yeah we can't get this right at present. It's a bug. The workaround would just be to re-annotate:

```python
class NotMyBaseClass(MyBaseClass):
    example: ClassVar[set[str]] = {"O", "M", "G", "Puppies!"}
```

**Comment by frenck** [comment]
> The workaround would just be to re-annotate:

Right... I understood that (and noticed), but that is kind of counterproductive / not a solution IMHO.

**Comment by charliermarsh** [comment]
Totally, I agree that it's not a good solution, but it's probably the best I can suggest immediately apart from "turn off the rule" :)

**Comment by frenck** [comment]
Re-annotation also overwrites the base. So, if the typing of the base class is changed, the children won't be updated (and might end up being incompatible).

So, this can actually cause bugs instead of preventing them. From that perspective, I would even argue the rule currently encourages something really bad.

../Frenck

**Comment by charliermarsh** [comment]
FWIW Mypy does warn on this, oddly Pyright does not:

```py
from typing import ClassVar

class MyBaseClass:
    example: ClassVar[set[str]]

class NotMyBaseClass(MyBaseClass):
    example: ClassVar[list[str]] = ["O", "M", "G", "Puppies!"]
```

**Comment by frenck** [comment]
👍 relied on Pyright in my previous answer, ok, that makes it a little less worse 😄

**Comment by charliermarsh** [comment]
Haha yeah. I'm not sure. I guess I need to think on it. Like, we could probably support this when both classes are defined in the same file, and soon (TM), when both classes are defined locally (as opposed to in third-party modules). But I'm wondering if the rule is too unreliable right now to be worth including in Ruff.

**Comment by charliermarsh** [comment]
In your case, do the base classes tend to be defined in the same file? In the same package (i.e., defined locally)? In third-party packages?

**Comment by frenck** [comment]
> In your case, do the base classes tend to be defined in the same file

Nope, they are not.

In the same package (i.e., defined locally)? In third-party packages?

Yes, they are defined locally, as in, not originating from a third-party package.

../Frenck

**Comment by nikkwong** [comment]
There's also a false negative, as no type error is raised when members that don't follow the types defined in a parent class:

```
class MyBaseClass:
    @abstractmethod
    def my_method(self, arg: str) -> Dict:
        pass

# no type errors raised:
class NotMyBaseClass(MyBaseClass):
    def my_method(self):
        return "not a dict"
```

**Comment by KerberosMorphy** [comment]
I was looking to report something similar so I'll join it here:

<img width="454" height="249" alt="Image" src="https://github.com/user-attachments/assets/4f71e571-5202-40ec-ac82-cb5cc36e4111" />

No error is detected in the `Test2` class for the `field` attribute.
